aiogram/dispatcher/requirement.py

```diff
@@ -0,0 +1,173 @@
+        err = context_manager.__exit__(type(e), e, None)
+        if not err:
+            raise e
+    else:
+        context_manager.__exit__(None, None, None)
+
+
+class Requirement(abc.ABC, Generic[T]):
+    """
+    Interface for all requirements
+    """
+
+    async def __call__(
+        self, cache_dict: Dict[CacheKeyType, Any], stack: AsyncExitStack, data: Dict[str, Any],
+    ) -> T:
+        raise NotImplementedError()
+
+
+class CallableRequirement(Requirement[T]):
+    __slots__ = "callable", "children", "cache_key", "use_cache", "generator_type"
+
+    def __init__(
+        self,
+        callable_: _RequiredCallback[T],
+        *,
+        cache_key: Optional[CacheKeyType] = None,
+        use_cache: bool = True,
+    ):
+        self.callable = callable_
+        self.generator_type = GeneratorKind.not_a_gen
+
+        if inspect.isasyncgenfunction(callable_):
+            self.generator_type = GeneratorKind.async_gen
+        elif inspect.isgeneratorfunction(callable_):
+            self.generator_type = GeneratorKind.plain_gen
+
+        self.cache_key = hash(callable_) if cache_key is None else cache_key
+        self.use_cache = use_cache
+        self.children = get_reqs_from_callable(callable_)
+
+        assert not (
+            set(inspect.signature(callable_).parameters) ^ set(self.children)
+        ), "Required can't manage callbacks with non `Requirement` parameters"
+
+    def filter_kwargs(self, data: Dict[str, Any]) -> Dict[str, Any]:
+        return {key: value for key, value in data.items() if key in self.children}
+
+    async def initialize_children(
+        self, data: Dict[str, Any], cache_dict: Dict[CacheKeyType, Any], stack: AsyncExitStack
+    ) -> None:
+        for req_id, req in self.children.items():
+            if isinstance(req, CachedRequirement):
+                data[req_id] = await req(data=data, cache_dict=cache_dict, stack=stack)
+                continue
+
+            if isinstance(req, CallableRequirement):
+                await req.initialize_children(data, cache_dict, stack)
+
+                if req.use_cache and req.cache_key in cache_dict:
+                    data[req_id] = cache_dict[req.cache_key]
+
+                else:
+                    data[req_id] = await initialize_callable_requirement(req, data, stack)
+
+                if req.use_cache:
+                    cache_dict[req.cache_key] = data[req_id]
+
+    async def __call__(
+        self, cache_dict: Dict[CacheKeyType, Any], stack: AsyncExitStack, data: Dict[str, Any],
+    ) -> T:
+        await self.initialize_children(data, cache_dict, stack)
+
+        if self.use_cache and self.cache_key in cache_dict:
+            return cast(T, cache_dict[self.cache_key])
+        else:
+            result = await initialize_callable_requirement(self, data, stack)
+            if self.use_cache:
+                cache_dict[self.cache_key] = result
+            return result
+
+
+class CachedRequirement(Requirement[T]):
```

Comment:
Cache mode can be controlled via init arguments without adding one more class?
Users can be confused for too much names of single feature

For example by `use_cache: bool = True`

or by `cache_key: Union[bool, CacheKeyType] = True`
in this case when value is True - generate key automatically, when False - don't use caching.

Reply:
You're right. However, keep in mind that we have bunch of inspections of `callable` in `CallableRequirement.__init__`, whereas `CachedRequirement` is just a cache-key holder.

I will definitely,  rethink design for requirements